filesource.go

```diff
@@ -216,48 +216,53 @@ func (s *FileSource) runMergeFile() error {
 	}
 }
 
-func (s *FileSource) streamIncomingFile(newIncomingFile *incomingBlocksFile, blocksStore dstore.Store) error {
-	defer func() {
-		close(newIncomingFile.blocks)
-	}()
-
-	atomic.AddInt64(&currentOpenFiles, 1)
-	s.logger.Debug("open files", zap.Int64("count", atomic.LoadInt64(&currentOpenFiles)), zap.String("filename", newIncomingFile.filename))
-	defer atomic.AddInt64(&currentOpenFiles, -1)
-
-	// FIXME: Eventually, RETRY for this given file.. and continue to write to `newIncomingFile`.
-	reader, err := blocksStore.OpenObject(context.Background(), newIncomingFile.filename)
-	if err != nil {
-		return fmt.Errorf("fetching %s from block store: %w", newIncomingFile.filename, err)
-	}
-	defer reader.Close()
+type retryableError struct {
+	error
+}
 
-	blockReader, err := s.blockReaderFactory.New(reader)
-	if err != nil {
-		return fmt.Errorf("unable to create block reader: %w", err)
+func (e retryableError) Error() string {
+	return e.error.Error()
+}
+func retryable(err error) error {
+	return retryableError{
+		error: err,
 	}
+}
+func isRetryable(err error) bool {
+	_, ok := err.(retryableError)
+	return ok
+}
 
+func (s *FileSource) streamReader(blockReader BlockReader, prevLastBlockRead BlockRef, output chan *PreprocessedBlock) (lastBlockRead BlockRef, err error) {
+	prevPassed := prevLastBlockRead == nil
 	for {
 		if s.IsTerminating() {
-			s.logger.Info("shutting down incoming batch file download", zap.String("filename", newIncomingFile.filename))
-			return nil
+			return
 		}
 
-		blk, err := blockReader.Read()
+		var blk *Block
+		blk, err = blockReader.Read()
 		if err != nil && err != io.EOF {
-			return fmt.Errorf("block reader failed: %w", err)
+			return lastBlockRead, retryable(err) // unexpected error
 		}
 
-		// EOF can happen with valid data, so let's skip if no block defined
 		if err == io.EOF && (blk == nil || blk.Num() == 0) {
-			return nil
+			return lastBlockRead, nil // EOF without data
```

Comment:
The previous comment seems a bit more descriptive IMO. What about
```suggestion
			return lastBlockRead, nil // EOF without a block read, no more data to consume
```